### DISCUSSION TITLE
SQLAlchemy connection pool does not exhaust

### DISCUSSION THREAD

**Discussion by frankshrestha** [sqlalchemy/sqlalchemy#7794] [Usage Questions] [answered]
From what I read in the documentation, [Connection Pool Configuration](https://docs.sqlalchemy.org/en/14/core/pooling.html#connection-pool-configuration),  
`The Engine returned by the create_engine() function in most cases has a QueuePool integrated, pre-configured with reasonable pooling defaults.`

I am using the `mysql` dialect. I have tried multiple drivers with it; `pymysql`, `mysqldb` and `mysqlconnector`. I am creating an engine like below.
```python
engine = create_engine(DB_URL, pool_size=5, poolclass=QueuePool, max_overflow=0)
```

When I've used a pool size of only `5`, why won't
```python
[engine.connect() for _ in range(20)]
```
exhaust the connection pool?

I also tried [Constructing a Pool](https://docs.sqlalchemy.org/en/14/core/pooling.html#constructing-a-pool) from the documentation.
```python
import sqlalchemy.pool as pool
import psycopg2

def getconn():
    c = psycopg2.connect(user='ed', host='127.0.0.1', dbname='test')
    return c

mypool = pool.QueuePool(getconn, max_overflow=0, pool_size=5)
```
The connection pool won't exhaust this way either.

What am I missing here? I wanted to ensure a connection per thread and check it the pool exhausts.

**Comment by zzzeek** [comment, marked as answer]
hi there -

I can't reproduce that.   Here's a python session, the code you have will pause for 30 seconds while it waits for the timeout and then throws an error

```
$ python
Python 3.10.0 (default, Nov  5 2021, 17:23:47) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sqlalchemy import create_engine
>>> from sqlalchemy.pool import QueuePool
>>> e = create_engine("mysql://scott:tiger@localhost/test", poolclass=QueuePool, pool_size=5, max_overflow=0)
>>> [e.connect() for _ in range(20)]             # it hangs for 30 seconds here
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 1, in <listcomp>
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/engine/base.py", line 2780, in connect
    return self._connection_cls(self)
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/engine/base.py", line 142, in __init__
    self._dbapi_connection = engine.raw_connection()
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/engine/base.py", line 2804, in raw_connection
    return self.pool.connect()
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/pool/base.py", line 388, in connect
    return _ConnectionFairy._checkout(self)
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/pool/base.py", line 1171, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/pool/base.py", line 644, in checkout
    rec = cast(_ConnectionRecord, pool._do_get())
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/pool/impl.py", line 156, in _do_get
    raise exc.TimeoutError(
sqlalchemy.exc.TimeoutError: QueuePool limit of size 5 overflow 0 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/20/3o7r)
```

can you try out the above as written?

**Reply by frankshrestha** [reply]
Hi,
Thanks for the response.
I tried the same thing with `pool_timeout=0` and it throws an exception. I would like to know if there is a way to make python script wait until the timeout.

**Reply by zzzeek** [reply]
The exception throw, assuming you are referring to `TimeoutError`, is normal.    Set the timeout to whatever number of seconds you desire and a script which calls upon engine.connect() will wait that number of seconds if no connections are available, before timing out, which is defined as the `TimeoutError` is thrown.

**Reply by frankshrestha** [reply]
Hi @zzzeek,
>Set the timeout to whatever number of seconds you desire and a script which calls upon `engine.connect()` will wait that number of seconds if no connections are available, before timing out, which is defined as the `TimeoutError` is thrown.

Tried it and it works. Thanks.
Previously, I was using threads where a thread would take one connection from the pool and surprisingly it did not wait for the timeout.
